File: src/Ray.ts
import { Matrix4x4 } from './Matrices';
import { Point, Vector } from './PointVector';

export class Ray {
  public readonly origin: Point;
  public readonly direction: Vector;

  public constructor(origin: Point, direction: Vector) {
    this.origin = origin;
    this.direction = direction;
  }

  public position(t: number): Point {
    return this.origin.add(this.direction.multiply(t));
  }

  public transform(matrix: Matrix4x4): Ray {
    const newOrigin = matrix.multiplyByPoint(this.origin);
    const newDirection = matrix.multiplyByVector(this.direction);
    return new Ray(newOrigin, newDirection);
  }
}

File: CSharp/RayTracerChallenge.Library/Ray.cs
﻿// ---------------------------------------------------------------------------------------------------------------------
// <copyright file="Ray.cs" company="Justin Rockwood">
//   Copyright (c) Justin Rockwood. All Rights Reserved. Licensed under the Apache License, Version 2.0. See
//   LICENSE.txt in the project root for license information.
// </copyright>
// ---------------------------------------------------------------------------------------------------------------------

namespace RayTracerChallenge.Library
{
    public sealed class Ray
    {
        //// ===========================================================================================================
        //// Constructors
        //// ===========================================================================================================

        public Ray(Point origin, Vector direction)
        {
            Origin = origin;
            Direction = direction;
        }

        //// ===========================================================================================================
        //// Properties
        //// ===========================================================================================================

        public Point Origin { get; }
        public Vector Direction { get; }

        //// ===========================================================================================================
        //// Methods
        //// ===========================================================================================================

        public Point PositionAt(double t)
        {
            return Origin + (Direction * t);
        }

        public Ray Transform(Matrix4x4 transform)
        {
            Point newOrigin = transform * Origin;
            Vector newDirection = transform * Direction;
            return new Ray(newOrigin, newDirection);
        }
    }
}
